runtime/gc_glue_java/FlattenedArrayObjectScanner.hpp

```diff
@@ -0,0 +1,279 @@
+		fprintf(stderr, "%.*s ", J9UTF8_LENGTH(name), J9UTF8_DATA(name));
+		fprintf(stderr, "(%p, %p) ", scanPtr, endPtr);
+		fprintf(stderr, "elementCount=%ld, ", sizeInElements);
+		fprintf(stderr, "elementSize=%ld, ", elementSize);
+		fprintf(stderr, "elementStride=%ld", elementStride);
+		fprintf(stderr, "\n");
+
+#if defined(OMR_GC_LEAF_BITS)
+		new(objectScanner) GC_FlattenedArrayObjectScanner(env, objectPtr, basePtr, limitPtr, scanPtr, endPtr, elementSize, elementStride, instanceDescription, leafDescription, flags);
+#else /* defined(OMR_GC_LEAF_BITS) */
+		new(objectScanner) GC_FlattenedArrayObjectScanner(env, objectPtr, basePtr, limitPtr, scanPtr, endPtr, elementSize, elementStride, instanceDescription, flags);
+#endif /* defined(OMR_GC_LEAF_BITS) */
+
+		objectScanner->initialize(env);
+		if (0 != startIndex) {
+			objectScanner->clearHeadObjectScanner();
+		}
+		return objectScanner;
+	}
+
+	MMINLINE uintptr_t getBytesRemaining() { return sizeof(fomrobject_t) * (_endPtr - _scanPtr); }
+
+	/**
+	 * @param env The scanning thread environment
+	 * @param allocSpace pointer to space within which the scanner should be instantiated (in-place)
+	 * @param splitAmount The maximum number of array elements to include
+	 * @return Pointer to split scanner in allocSpace
+	 */
+	GC_IndexableObjectScanner *
+	splitTo(MM_EnvironmentBase *env, void *allocSpace, uintptr_t splitAmount)
+	{
+		Assert_MM_unimplemented();
+	}
+
+	/**
+	 * Return base pointer and slot bit map for next block of contiguous slots to be scanned. The
+	 * base pointer must be fomrobject_t-aligned. Bits in the bit map are scanned in order of
+	 * increasing significance, and the least significant bit maps to the slot at the returned
+	 * base pointer.
+	 *
+	 * @param[out] scanMap the bit map for the slots contiguous with the returned base pointer
+	 * @param[out] hasNextSlotMap set this to true if this method should be called again, false if this map is known to be last
+	 * @return a pointer to the first slot mapped by the least significant bit of the map, or NULL if no more slots
+	 */
+	virtual fomrobject_t *
+	getNextSlotMap(uintptr_t *slotMap, bool *hasNextSlotMap)
+	{
+		fomrobject_t *result = GC_HeadlessMixedObjectScanner::getNextSlotMap(slotMap, hasNextSlotMap);
+		if (result != NULL) {
+			/* Ignore hasNextSlotMap from HeadLess, we want to always report that there is another element */
+			*hasNextSlotMap = true;
+			return result;
+		}
+
+		// No more slots in the current element, get the next element of the array
+		result = _indexableScanner.next();
+		if (result == NULL) {
+			// There are no elements in the array
+			*hasNextSlotMap = false;
+			return result;
+		}
+
+		// Reset the scanner to start scanning the next element
+		_mapPtr = result;
+		_endPtr = (fomrobject_t *)((uint8_t*)_mapPtr + _elementSizeWithoutPadding);
+		GC_HeadlessMixedObjectScanner::initialize(_env, _descriptionBasePtr);
+		*hasNextSlotMap = true;
+		return result;
+	}
```

Comment:
Is this more readable (if I got it right):


		fomrobject_t *result = GC_HeadlessMixedObjectScanner::getNextSlotMap(slotMap, hasNextSlotMap);
		
		/* Ignore hasNextSlotMap from HeadLess, we want to always report that there is another slotMap,
		  but indexable scanner may tell us otherwise (if this is really the last element) */
		*hasNextSlotMap = true;
			
		if (NULL == result) {
			// No more slots in the current element, get the next element of the array
			result = _indexableScanner.next();
			if (NULL == result) {
				// There are no elements in the array
				*hasNextSlotMap = false;
			}  else {			
				// Reset the scanner to start scanning the next element
				_mapPtr = result;
				_endPtr = (fomrobject_t *)((uint8_t*)_mapPtr + _elementSizeWithoutPadding);
				GC_HeadlessMixedObjectScanner::initialize(_env, _descriptionBasePtr);
			}
		}
		
		return result;

Reply:
Yeah, I think this is more readable. The multiple returns in the original code were pretty confusing. I'll update it.

Reply:
Done.